redis_state_sync.go

```diff
@@ -108,3 +131,60 @@ func (r *RedisStateSync) GetStartSequence(shardID *string) *string {
 		return nil
 	}
 }
+
+func (r *RedisStateSync) TryAcquire(shardID *string) error {
+	conn := r.pool.Get()
+	defer conn.Close()
+	if _, exists := r.acquired[*shardID]; exists {
+		return errors.New("Lock already acquired by this process")
+	}
+	res, err := conn.Do("SET", r.redisPrefix+".lock."+*shardID, r.lock, "PX", r.alivePeriod/time.Millisecond, "NX")
+	if err != nil {
+		return err
+	}
+	if res != "OK" {
+		return errors.New("Failed to acquire lock")
+	}
+	r.acquired[*shardID] = Unit{}
+	return nil
+}
+
+func (r *RedisStateSync) Reacquire() {
+	conn := r.pool.Get()
+	defer conn.Close()
+	for shardID := range r.acquired {
+		res, err := conn.Do("PEXPIRE", r.redisPrefix+".lock."+shardID, r.alivePeriod/time.Millisecond, "NX")
+		if err != nil || res != "OK" {
+			if err != nil {
+				r.recs <- &KinesisRecord{
```

Comment:
Don't have to fix this now, but it's a little weird that we're using the kmiesis records channel as the error channel also. May want to consider separating these in the future.